File: 백준/Silver/1326. 폴짝폴짝/폴짝폴짝.py
import sys
from collections import deque
def input(): return sys.stdin.readline().rstrip()

n=int(input())
nb=list(map(int,input().split()))
a,b=map(lambda x: int(x)-1,input().split())

vt=[-1]*n
dq=deque()

dq.append((a,nb[a]))
vt[a]=0
while dq:
    idx,x=dq.popleft()
    if idx==b: break
    start=idx%x
    for i in range(start,n,x):
        if vt[i]<0:
            vt[i]=vt[idx]+1
            dq.append((i,nb[i]))

print(vt[b])

File: 백준/Silver/1326. 폴짝폴짝/폴짝폴짝.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws Exception{
        BufferedReader I=new BufferedReader(new InputStreamReader(System.in));

        int n=Integer.parseInt(I.readLine());
        int[] nb=new int[n];
        StringTokenizer st=new StringTokenizer(I.readLine());
        for(int i=0;i<n;i++) nb[i]=Integer.parseInt(st.nextToken());

        st=new StringTokenizer(I.readLine());
        int a=Integer.parseInt(st.nextToken())-1,b=Integer.parseInt(st.nextToken())-1;

        int[] vt=new int[n];
        Arrays.fill(vt,-1);

        vt[a]=0;

        Deque<Integer> dq=new ArrayDeque<>();
        dq.offer(a);
        dq.offer(nb[a]);

        while(!dq.isEmpty()){
            int idx=dq.pollFirst(),x=dq.pollFirst();
            if(idx==b) break;

            for(int start=idx%x;start<n;start+=x){
                
                if(vt[start]<0){
                    vt[start]=vt[idx]+1;
                    dq.offer(start); dq.offer(nb[start]);
                }
            }
        }
        
        System.out.print(vt[b]);
    }
}